### DISCUSSION TITLE
New-Function Set-Function Get-Function

### DISCUSSION THREAD

**Discussion by Molochnikov** [PowerShell/PowerShell#16881] [General]
### Summary of the new feature / enhancement

Why the hell in 2022 to create function in parent scope i need to do this:
```
 function b() {
    function a() {
        function d() {
            $f = (new-item -path ("function:test") -value {Write-Host Hello!});
            $context = $host.runspace.gettype().getmethod("get_ExecutionContext",([System.Reflection.BindingFlags]::NonPublic -bor [System.Reflection.BindingFlags]::Instance)).Invoke($host.runspace,$null)
            $ss = $context.gettype().getmethod("get_EngineSessionState",([System.Reflection.BindingFlags]::NonPublic -bor [System.Reflection.BindingFlags]::Instance)).Invoke($context,$null)
            $scope = $ss.gettype().getmethods(([System.Reflection.BindingFlags]::NonPublic -bor [System.Reflection.BindingFlags]::Instance)) | Where-Object {$_.name -eq "GetScopeByID"} | ForEach-Object {try {$_.Invoke($ss,1)} catch {}}
            $funct = $scope.gettype().getmethod("get_FunctionTable",([System.Reflection.BindingFlags]::NonPublic -bor [System.Reflection.BindingFlags]::Instance)).Invoke($scope,$null)
            $Funct.Add('SayHello', $f)
        }
        d
        SayHello #good
    }
    a
    SayHello #error
 }
 b
 SayHello #error
```

I see that you like variables much more than functions:
`Set-Variable -Name canCreateVariableInParentScope -Value $true -Scope 1`

I'm serious. You all racists...

Please add that fuckin `New-Set-Get-Function` commands with `-Scope` parameter!

**Comment by jhoneill** [comment]
@Molochnikov  
1. More respectful language is expected here. 
2. This may give you the clue you seem to be missing. 
```
function foo {
  write-host "creating bar" 
  function global:bar {"I'm Global"}
}

> bar
bar : The term 'bar' is not recognized as the name of a cmdlet...

> foo
creating bar

> bar
I'm Global
```
3.  A quick test showed this works for the script: scope but not for numbered scopes.  For this you pass the script block to run as a variable and run it with the call operator 
```
function foo2 {
  write-host "creating sb" 
  Set-Variable -name sb -Scope 1 -value {
     param($what) 
    "I'm in $what"
  }
}
function wibble {
  foo2
  & $sb -what "wibble" 
}
```
4. As an alternative a function which calls another an wants it to set things in its scope can DOT source the function it calls
```
function foo3 {
  function barbar {
    param ($What) 
    "What scope"
  }
}

function fuz {
  . foo3 
  barbar
}

> fuz
What scope
```
5. Changing the variables in the parent scope is considered "Bad but occasionally necessary".  Values inherit - what applies in a parent scope applies in a child scope - parameters are passed by reference so changing a member of passed variable or an inherited variable  changes that member everywhere, but assigning a value to a variable name creates a new variable  in the current scope (so setting the value of a number or a string won't pass upwards). 
```

 function eckie {
$a.x = "99"
$b =  @{x="99";"y"=100}
}

> $a = @{x=1;y=2} ;$b = @{x=1;y=2}
> $a,$b

Name                           Value                                                                                                                                                              
----                           -----                                                                                                                                                              
y                              2                                                                                                                                                                  
x                              1                                                                                                                                                                  
y                              2                                                                                                                                                                  
x                              1                                                                                                                                                                  

> eckie

> $a,$b

Name                           Value                                                                                                                                                              
----                           -----                                                                                                                                                              
y                              2                                                                                                                                                                  
x                              99             
y                              2                                                                                                                                                                  
x                              1            
 ##### The child scope changed A but created a new B 
```

6. Only rarely does a _variable_ need to be set in a parent scope instead of returned by a function and it's seen as a bad smell.   Do you have a real-world use for doing this with functions., instead of dot sourcing or handing back a script block obviously  `name -param` is slightly neater than `& $name -param`  but from a good coding practice point of view making it super clear that something being executed was the result of executing something else does seem better

**Comment by Molochnikov** [comment]
My statement is: There is no standart way to operate functions in **parent** scopes **from child** scope. Documentation **about_Scopes** states that:

> You can declare variables, aliases, **functions**, and PowerShell drives in a scope **outside** of the **current** scope.

There is no excuses to do it for variables without doing the same for functions.

**Comment by Molochnikov** [comment]
One practical application will be pushing closures to the parent scope in functional programming:

 ```
       function closure($name) {
            $v = 'some variable';
            $f = (new-item -path ("function:" + $name) -value {return $v}.GetNewClosure());
            $context = $host.runspace.gettype().getmethod("get_ExecutionContext",([System.Reflection.BindingFlags]::NonPublic -bor [System.Reflection.BindingFlags]::Instance)).Invoke($host.runspace,$null)
            $ss = $context.gettype().getmethod("get_EngineSessionState",([System.Reflection.BindingFlags]::NonPublic -bor [System.Reflection.BindingFlags]::Instance)).Invoke($context,$null)
            $scope = $ss.gettype().getmethods(([System.Reflection.BindingFlags]::NonPublic -bor [System.Reflection.BindingFlags]::Instance)) | Where-Object {$_.name -eq "GetScopeByID"} | ForEach-Object {try {$_.Invoke($ss,1)} catch {}}
            $funct = $scope.gettype().getmethod("get_FunctionTable",([System.Reflection.BindingFlags]::NonPublic -bor [System.Reflection.BindingFlags]::Instance)).Invoke($scope,$null)
            $Funct.Add($name, $f)
        }

        closure 'test'
        test

```

**Comment by 237dmitry** [comment]
> One practical application will be pushing closures

Functions can be anonymous.
```powershell
function closure
{
    param ([string] $n)
    return ({ param ([string] $m) return $n + $m }.GetNewClosure())
}                              
$a = closure 'test + '         
$a.Invoke('one more test')
& $a 'the second test' 
```

**Comment by jhoneill** [comment]
That text does not appear in the 7.2.1 help. 

Functions in an ancestor scope can be called from any level of descendant 
Any scope can define or redefine a function in the global scope
You are complaining about a "worst practice" being blocked - that if Scope X defines a function then descendant scope Y cannot change that function's behaviour when called later from X 

@237dmitry  has just what I was going to post posted what I was going to with one change which I should have remembered when I gave example `#3` in the first post.  When passed about function can be anonymous but they don't need to be.  

```
function closure {
    param ([string] $n)
    return ({ param ([string] $m) return $n + $m }.GetNewClosure())
}                              

$function:Test =  closure 'test + ' 
Test "blah blah"
``` 

A function should not use variables in another scope as its parameters or for its results.

**Comment by Molochnikov** [comment]
https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_scopes?view=powershell-7.3

> You can declare variables, aliases, functions, and PowerShell drives in a scope outside of the current scope.

> You are complaining about a "worst practice" being blocked - that if Scope X defines a function then descendant scope Y cannot change that function's behaviour when called later from X

It may be "worst practice" case, but it is a case. Reflection is a bad practice too. Very few programmers using Reflection.Emit, but dotNet library contains it.

For functional/symbolic programming it is a case. If you can break things by changing outer scope variables why you cannot break things by changing outer scopes functions, aliases, drives? What bad in creating function in outer scope (and only upper 1,2,3... scopes) if outer scope will never contains a code that calls it?

<blockquote><img src="https://docs.microsoft.com/media/logos/logo-powershell-social.png" width="48" align="right"><div><strong><a href="https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_scopes">about Scopes - PowerShell</a></strong></div><div>Explains the concept of scope in PowerShell and shows how to set and change the scope of elements.</div></blockquote>

**Comment by Molochnikov** [comment]
What if i don't know which functions child scope will change?

Compare the code:
```
function closure1 {
    param ([string] $n)
    function closure2 {
        param ([string] $n)
        function closure3 {
            param ([string] $n)
            return ({ param ([string] $m) return $n + 1 }.GetNewClosure())
        }
        $function:Test =  closure3 'test + '
        Test "blah blah"
        return @(({ param ([string] $m) return $n + 2 }.GetNewClosure()),$function:Test)
    }
    $res = closure2 'test + '
    $function:Test =  $res[1]
    $function:Test2 = $res[0]
    Test "blah blah"
    Test2 "blah blah"
    return @(({ param ([string] $m) return $n + 3 }.GetNewClosure()), $function:Test, $function:Test2)
}                              

...


function closure1 {
    param ([string] $n)
    function closure2 {
        param ([string] $n)
        function closure3 {
            param ([string] $n)
            set-function -scope 2 -name Test -value ({ param ([string] $m) return $n + 1 }.GetNewClosure())
        }
        Test "blah blah"
        Test13424 
        Test "blah blah"
        Test322
        set-function -scope 1 -name Test2 -value ({ param ([string] $m) return $n + 2 }.GetNewClosure())
    }
    Test "blah blah"
    Test2 "blah blah"
   Test3473842
Test3473843
Test3473844
    return ({ param ([string] $m) return $n + 3 }.GetNewClosure())
}

...
```

**Comment by jhoneill** [comment]
Back when I was doing my Computer Science degree (which, BTW did involve some functional programming), if I had turned in an assignment that looked like either of those it would have got a D or D- (assuming it worked).  I don't think it is working as intended because 
```
 Test "blah blah"
  return @(({ param ([string] $m) return $n + 2 }.GetNewClosure()),$function:Test)
```
returns 3 things and the part which calls closure2 is expecting two closures as the first and second assuming you wanted the whole that result its easier to write
```
function closure3 {
    param ([string] $n)
    {param ([string] $m) $n + 1 }.GetNewClosure()
}
function closure2 {
    param ([string] $n)
    Test "Ignored Param from C2"
    {param ([string] $m) $n + 2 }.GetNewClosure() 
} 
$var                    = 'test + '
$function:Test          =  closure3 $var
$c2Out, $function:Test2 =  closure2 $var
$c2out 
Test  "Ignored"
Test2 "ignored2"
````
I.e. if you want the result of closure 3 inside and outside closure 2, define it and call it outside closure 2.  Instead of defining things at the highest scope they will be needed and letting them inherit down, you're trying to define them in the scope that where will call them first and then force them up to where they will be needed later. In the version above the call to run `Test "ignored"` can be moved before `closeure2` is invoked. In your version to changing the sequence means changing the structure.

I'd love to see a real world case where anyone actually wants to write anything like your example.

**Comment by Molochnikov** [comment]
Just want to post it here:
#### New developing simplifyed languages:

> Hello, can you add base feature in your language to work with functions? Your docs saying that your language can do it.

Wow, we don't have this base feature! Yes, of course, will be on the next release!

#### Dying overburdened languages with millions ways to do the same thing but don't have a base feature:

Who are you? What you do here? What you want? Why you want this? I don't use it. Im professional. Docs are true. It is not the base feature. Get out of here!

**Reply by doctordns** [reply]
That is both inaccurate and unhelpful. The posters here care about PowerShell - and we're all keen to add new features. But only where those features make sense. So far in this thread, I have not seen a decent use case for the change being proposed.

**Comment by Molochnikov** [comment]
Its funny that you can declare/change function in **global** scope from any scope, but cannot do it for **parent** scope from **child scope.** For me **thats is the REAL bad practice**. Not theoretical imaginary bad practice.

**Comment by jhoneill** [comment]
@Molochnikov  

in the real world.... 

When a feature is missing and a significant number of people want it, or an argument is made to the people who would need to work on it that they would want it if were there, then that feature **gets  the chance to compete for the limited resources available for its implementation**.   There are standard reasons given to say no. "No-one wants it.", "People want it, but it goes against some principle", "The resource required isn't justified",  

When someone turns up demanding something which no-one else has ever asked for, using the "F word" and calling people racist (seriously, what ? )  most people won't give it any time.  And your response to explanations doesn't encourage people to support you. 

I'm not going to propose _function_ equivalents of the _variable_ commands in the cmdlet working group. The TLDR version is ** it's not new , not good and not wanted .** 
1. The premise that things can be done with variables but not functions is simply wrong. Function are specialized variables holding script blocks.  You can run a script block from a normal variable with the call operator `&`  or put one into a `$function:variable`, instead of using the `function` keyword.  Your proposal is "different syntax for existing functionality". 
2. Your latest contention that setting things in Global scope is worse than setting in a parent scope is also wrong. From a strict computer science point of view, even the inheritance we have is bad - values _should_ pass into function as parameters and out as return values. In reality things like the path, the machine name, CurrentProcessID, or working directory are variables which apply globally access to them is simplest via variables   `Connect-ToService` functions might use "ServiceConnection" as a well known name for something applicable to code it can't know about so that _is_ naturally global. Global variables are only bad when used for things which are not global in nature.  There are cases like the `WarningVariable` parameter where a parameter specifies a location to put something which isn't to be returned as a result, but when code in a child scope _takes it upon itself to change anything_ in a scope which called that's bad code.  If doing a search and replace for a variable name changes the effect of calling something - that's bad code - hence my comments on getting a low grade for code which does just that.  All this means that  although `Set-Variable` can be abused it is _necessary_, but anything which works in a similar way needs to show it is necessary and can't realistically be done any other way:  this does not. 
3. Demand? This is the first request and has 2 down votes and zero up votes. I'm sure it would save you time, but the total time you would save may never repay the time taken to add the feature. 

If you persuade me (or someone else) that is good and/or widely wanted you might make some progress. But you're not winning people over. 

If you want to contrast mature and immature languages; Immature ones tend to go "Ooh something new let's do it".  And that contributes to many of them not making it to maturity. Mature ones go "We've seen it, solved it like this, that was good enough for the the one person who cared then, isn't that good enough for you?".

**Comment by doctordns** [comment]
Having read this thread again today, this does not seem to me to be a problem we need to solve today, especially as "fixing" it could well be a  breaking change.  I've seen no community support for such a change either.

As ever, the OP is more than welcome to fork this repo, make the change and let us evaluate it.  But I recognise that that is not a particularly easy task

I suggest we close this as  'won't fix' and 'by design'

**Comment by mklement0** [comment]
Let me attempt a summary, adapted from [this Stack Overflow answer](https://stackoverflow.com/a/79193354/45375):

* While you can target other scopes for creating functions via _scope modifiers_ (e.g., `function global:Foo { 'hi!' }`), there is *no* way to specify a _relative_ scope by way of a scope modifier, as would be needed if you wanted to define a function in the _caller's_ scope - whatever scope the caller lives in.

   * _Relative_ scopes, e.g. `1` to refer to the _parent_ scope, are only supported via the `-Scope` parameter of *entity-specific* cmdlets, such as [`Set-Variable`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/set-variable) and [`Set-Alias`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/set-alias); there is _no_ <del>`Set-Function`</del> cmdlet.

     * Conceivably, however, even the generic [provider](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Providers) cmdlets, such as `New-Item` and `Set-Content`, could be extended to provide entity-specific _dynamic_ parameters (an existing example of such parameters are the `-File` and `-Directory` switches that are specific to the `FileSystem` provider).

The following is a streamlined version of the code in the initial post, provided in the form of a `New-FunctionInCallerScope` function.
It targets the parent scope (the equivalent of `-Scope 1`) if defined outside a module, and an _outside_ caller's current scope, if defined in a module.

Perhaps needless to say, this solution is still _suboptimal_, due to its complexity and, primarily, due to use of non-public APIs.

```powershell
function New-FunctionInCallerScope {
  param(
    [Parameter(Mandatory)] [string] $FunctionName,
    [Parameter(Mandatory)] [scriptblock] $Body
  )

  # Binding flags for finding non-public instance members via reflection.
  $bindingFlags = [System.Reflection.BindingFlags] 'NonPublic, Instance'
  
  # Get the caller's internal session state, via $PSCmdlet.SessionState
  $sessionStateInternal = [System.Management.Automation.SessionState].
                            GetMethod('get_Internal', $bindingFlags).
                            Invoke($PSCmdlet.SessionState, $null)
  # Determine the scope index relative to $PSCmdlet.SessionState[Internal],
  # which is the *caller's* session state:
  # * If this function is inside a module, it lives in a different "session state"
  #   (scope domain), so use the caller's *current* scope (0) 
  #   (the caller is assumed to be an *outside* caller).
  # * If this function is a non-module function, it runs in the same "session state"
  #   as the caller, but in a *child* scope (unless dot-sourced), so use the parent
  #   scope (1).
  $callerScopeIndex = if ($MyInvocation.MyCommand.Module) { 0 } else { 1 }
  # Retrieve the target scope.
  $scope = $sessionStateInternal.GetType().
             GetMethod('GetScopeByID', $bindingFlags, [type[]] [int]).
             Invoke($sessionStateInternal, $callerScopeIndex)
  # Get the scope's function table. 
  $funcTable = $scope.GetType().
                 GetMethod('get_FunctionTable', $bindingFlags).
                 Invoke($scope, $null)
  # Create / update the function with the given name and script block.
  # Note: A [System.Management.Automation.FunctionInfo] instance must be assigned, which is
  #       is obtained via an aux. transitory, scope-local function whose name doesn't matter.
  $funcTable[$FunctionName] = New-Item Function:UnusedName -Value $Body
}

# Sample calls:

# Define function foo in the caller's scope.
New-FunctionInCallerScope foo { 'hi' }

# The function is now visible in this scope (and its descendants), 
# but not in any parent scope.
foo
```

**Comment by nightroman** [comment]
I miss the suggested feature. I had use cases for creating functions in the parent (or particular) scope. I am a little bit surprised by the asymmetric PowerShell design with respect to functions. As if functions are not the same class citizens as aliases and variables. Maybe there are some solid tech reasons for this though, I cannot tell.
